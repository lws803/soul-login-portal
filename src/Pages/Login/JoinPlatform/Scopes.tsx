import React from 'react';
import { Box, Divider, Text, VStack } from '@chakra-ui/react';
import ScopeRow from './Scopes/ScoprRow';

export default function Scopes() {
  return (
    <Box
      borderRadius="md"
      borderStyle="solid"
      borderColor="soul.mutedGrey"
      borderWidth="1px"
      padding="1rem"
    >
      <VStack alignItems="left">
        <Text>
          Joining this platform will allow the platform and it&apos;s developers
          to:
        </Text>
        <Divider />
        <VStack alignItems="left" overflowY="auto" maxHeight="30vh">
          <ScopeRow title="Access your account information." />
          <ScopeRow title="Modify your account." />
          <ScopeRow title="Access your user connections." />
          <ScopeRow title="Modify your user connections." />
          <ScopeRow title="Access your membership on this platform." />
          <ScopeRow title="Modify your membership on this platform." />
        </VStack>
      </VStack>
    </Box>
  );
}
