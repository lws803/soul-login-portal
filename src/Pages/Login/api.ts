import axios from 'axios';
import axiosRetry from 'axios-retry';

import { BASE_URL } from 'src/constants';
import { ApiError } from 'src/shared/apiTypes';

axiosRetry(axios, { retryDelay: axiosRetry.exponentialDelay, retries: 3 });

export const loginWithPlatform = async ({
  platformId,
  callback,
  values,
  state,
  codeChallenge,
}: {
  platformId: number;
  callback: string;
  values: {
    email: string;
    password: string;
  };
  state: string;
  codeChallenge: string;
}): Promise<{
  data: LoginWithPlatformResponse | null;
  error: ApiError | null;
}> => {
  try {
    const { data } = await axios.post<LoginWithPlatformResponse>(
      `${BASE_URL}/v1/auth/code`,
      values,
      {
        params: {
          client_id: platformId,
          redirect_uri: callback,
          state,
          code_challenge: codeChallenge,
        },
      },
    );
    return { data, error: null };
  } catch ({ response: { data } }) {
    return { data: null, error: data as ApiError };
  }
};

type LoginWithPlatformResponse = {
  code: string;
  state: string;
};

export const login = async ({
  email,
  password,
}: {
  email: string;
  password: string;
}) => {
  try {
    const { data } = await axios.post<LoginResponse>(
      `${BASE_URL}/v1/auth/login`,
      { email, password },
    );
    return { data, error: null };
  } catch ({ response: { data } }) {
    return { data: null, error: data as ApiError };
  }
};

type LoginResponse = {
  access_token: string;
};

export const joinPlatformAndLogin = async ({
  platformId,
  callback,
  email,
  password,
  accessToken,
  state,
  codeChallenge,
}: {
  platformId: number;
  callback: string;
  email: string;
  password: string;
  accessToken: string;
  state: string;
  codeChallenge: string;
}) => {
  try {
    await axios.post(`${BASE_URL}/v1/platforms/${platformId}/join`, undefined, {
      headers: { Authorization: `Bearer ${accessToken}` },
    });

    const { data: codeData } = await axios.post<LoginWithPlatformResponse>(
      `${BASE_URL}/v1/auth/code`,
      { email, password },
      {
        params: {
          platform_id: platformId,
          redirect_uri: callback,
          state,
          code_challenge: codeChallenge,
        },
      },
    );
    return { data: codeData, error: null };
  } catch ({ response: { data } }) {
    return { data: null, error: data as ApiError };
  }
};
