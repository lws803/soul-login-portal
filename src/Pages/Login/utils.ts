export const redirectToCallback = ({
  callback,
  code,
  state,
}: {
  callback: string;
  code: string;
  state: string;
}) => {
  if (typeof window !== 'undefined') {
    window.open(`${callback}?code=${code}&state=${state}`, '_self');
  } else {
    throw new Error("Can't redirect to callback.");
  }
};
