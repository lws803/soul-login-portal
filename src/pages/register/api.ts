import axios from 'axios';

import { BASE_URL } from '../../constants';
import { ApiError } from '../../shared/api-types';

export const register = async ({
  values,
}: {
  values: { email: string; password: string; username: string };
}) => {
  try {
    const { data } = await axios.post<Response>(`${BASE_URL}/v1/users`, values);
    return { data, error: null };
  } catch ({ response: { data, status } }) {
    return {
      data: null,
      error: { data: data as ApiError, status: status as number },
    };
  }
};

type Response = {
  id?: number;
};
