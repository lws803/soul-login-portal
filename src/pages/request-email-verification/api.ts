import axios from 'axios';

import { BASE_URL } from '../../constants';
import { ApiError } from '../../shared/api-types';

export const sendPasswordReset = async ({ email }: { email: string }) => {
  try {
    await axios.post(
      `${BASE_URL}/v1/users/request-password-reset-token?email=${email}`,
    );
    return { error: null };
  } catch ({ response: { data } }) {
    return { error: data as ApiError };
  }
};
