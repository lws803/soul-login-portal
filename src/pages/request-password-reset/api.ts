import axios from 'axios';

import { BASE_URL } from '../../constants';

import { ApiError } from '../../shared/api-types';

export const sendEmailConfirmation = async ({ email }: { email: string }) => {
  try {
    await axios.post(
      `${BASE_URL}/v1/users/resend-confirmation-token?email=${email}`,
    );
    return { error: null };
  } catch ({ response: { data } }) {
    return { error: data as ApiError };
  }
};
