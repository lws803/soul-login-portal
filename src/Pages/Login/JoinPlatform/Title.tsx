import { CheckCircleIcon } from '@chakra-ui/icons';
import { HStack, Tooltip, Text } from '@chakra-ui/react';
import { Platform } from '../api';

export default function Title({ platform }: Props) {
  return (
    <HStack justifyContent="center" w={['90vw', '90vw', '500px', '500px']}>
      <Text
        fontSize="3xl"
        whiteSpace="nowrap"
        overflow="hidden"
        textOverflow="ellipsis"
        fontWeight="bold"
        flexShrink={1}
      >
        {platform.name}
      </Text>
      {platform.isVerified && (
        <Tooltip label="This platform is verified" flexShrink={0}>
          <span>
            <CheckCircleIcon
              color="soul.green.200"
              aria-label="Verified marker"
            />
          </span>
        </Tooltip>
      )}
    </HStack>
  );
}

type Props = {
  platform: Platform;
};
