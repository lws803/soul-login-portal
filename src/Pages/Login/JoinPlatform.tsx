import { useEffect, useState } from 'react';
import {
  Center,
  HStack,
  Spinner,
  Text,
  Tooltip,
  VStack,
} from '@chakra-ui/react';
import { CheckCircleIcon } from '@chakra-ui/icons';

import FancyButton from 'src/components/FancyButton';

import {
  login,
  joinPlatformAndLogin,
  getPlatformDetails,
  Platform,
} from './api';
import { redirectToCallback } from './utils';

export default function JoinPlatform({
  email,
  password,
  platformId,
  callback,
  setErrors,
  state,
  codeChallenge,
  ...props
}: Props) {
  const [accessToken, setAccessToken] = useState<string>();
  const [platform, setPlatform] = useState<Platform>();
  const [isJoining, setIsJoining] = useState(false);

  useEffect(() => {
    const preLogin = async () => {
      const { data, error } = await login({ email, password });
      if (error) {
        setErrors([error.message]);
      }
      if (data) {
        setAccessToken(data.access_token);
      }
    };
    preLogin();
  }, [email, password, setErrors]);

  useEffect(() => {
    const getPlatform = async () => {
      const { data, error } = await getPlatformDetails({ platformId });
      if (error) {
        setErrors([error.message]);
      }
      if (data) {
        setPlatform(data);
      }
    };
    getPlatform();
  }, [platformId]);

  const joinPlatform = async () => {
    setIsJoining(true);
    const { data, error } = await joinPlatformAndLogin({
      platformId,
      callback,
      email,
      password,
      state,
      accessToken: accessToken!,
      codeChallenge,
    });
    if (error) {
      setErrors([error.message]);
    }
    if (data) {
      redirectToCallback({ code: data.code, state: data.state, callback });
    }
    setIsJoining(false);
  };

  if (!platform) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  return (
    <VStack spacing="32px" {...props}>
      <VStack justifyContent="center">
        <HStack justifyContent="center" w={['90vw', '90vw', '500px', '500px']}>
          <Text
            fontSize="3xl"
            whiteSpace="nowrap"
            overflow="hidden"
            textOverflow="ellipsis"
            fontWeight="bold"
            flexShrink={1}
          >
            {platform.name}
          </Text>
          {platform.isVerified && (
            <Tooltip label="This platform is verified" flexShrink={0}>
              <span>
                <CheckCircleIcon
                  color="soul.green.200"
                  aria-label="Verified marker"
                />
              </span>
            </Tooltip>
          )}
        </HStack>
        <Text
          color="soul.mutedGrey"
          whiteSpace="nowrap"
          overflow="hidden"
          textOverflow="ellipsis"
          w={['90vw', '90vw', '500px', '500px']}
          textAlign="center"
        >
          {platform.nameHandle}
        </Text>
      </VStack>
      <FancyButton
        onClick={joinPlatform}
        isLoading={!accessToken || isJoining}
        aria-label="Join platform button"
      >
        Join Platform!
      </FancyButton>
    </VStack>
  );
}

type Props = {
  email: string;
  password: string;
  platformId: number;
  callback: string;
  setErrors: (error: string[]) => void;
  state: string;
  codeChallenge: string;
} & React.ComponentProps<typeof VStack>;
