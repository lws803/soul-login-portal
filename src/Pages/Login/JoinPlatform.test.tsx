import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';

import JoinPlatform from './JoinPlatform';
import * as api from './api';
import * as utils from './utils';

describe(JoinPlatform, () => {
  const props: React.ComponentProps<typeof JoinPlatform> = {
    setErrors: jest.fn(),
    email: 'TEST@GMAIL.COM',
    password: 'PASSWORD',
    platformId: 1,
    callback: 'https://www.example.com',
    state: 'STATE',
    codeChallenge: 'CODE_CHALLENGE',
  };

  beforeEach(() => {
    jest
      .spyOn(api, 'login')
      .mockResolvedValue({ data: { access_token: 'TOKEN' }, error: null });
    jest.spyOn(api, 'getPlatformDetails').mockResolvedValue({
      data: { nameHandle: 'TEST#1', name: 'TEST', isVerified: true },
      error: null,
    });
  });

  it('renders', async () => {
    const { container, findByText } = render(<JoinPlatform {...props} />);

    await findByText('Join Platform!');
    expect(container).toMatchSnapshot();
  });

  it('joins platform when clicked', async () => {
    const joinPlatformAndLogin = jest
      .spyOn(api, 'joinPlatformAndLogin')
      .mockResolvedValue({
        data: { code: 'CODE', state: 'STATE' },
        error: null,
      });

    jest.spyOn(utils, 'redirectToCallback').mockImplementation();

    const { findByText } = render(<JoinPlatform {...props} />);

    fireEvent.click(await findByText('Join Platform!'));

    await waitFor(() => {
      expect(joinPlatformAndLogin).toHaveBeenCalledWith({
        accessToken: 'TOKEN',
        callback: 'https://www.example.com',
        email: 'TEST@GMAIL.COM',
        password: 'PASSWORD',
        platformId: 1,
        state: 'STATE',
        codeChallenge: 'CODE_CHALLENGE',
      });

      expect(utils.redirectToCallback).toHaveBeenCalledWith({
        callback: 'https://www.example.com',
        code: 'CODE',
        state: 'STATE',
      });
    });
  });
});
