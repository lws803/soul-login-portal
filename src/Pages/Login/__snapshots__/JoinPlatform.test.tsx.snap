// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinPlatform renders 1`] = `
<div>
  <div
    class="chakra-stack css-dvtbfj"
  >
    <div
      class="chakra-stack css-1jxjhsf"
    >
      <div
        class="chakra-stack css-1n382hm"
      >
        <p
          class="chakra-text css-1n5khlh"
        >
          TEST
        </p>
        <span>
          <svg
            aria-label="Verified marker"
            class="chakra-icon css-1rcwe32"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,0A12,12,0,1,0,24,12,12.014,12.014,0,0,0,12,0Zm6.927,8.2-6.845,9.289a1.011,1.011,0,0,1-1.43.188L5.764,13.769a1,1,0,1,1,1.25-1.562l4.076,3.261,6.227-8.451A1,1,0,1,1,18.927,8.2Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </div>
      <p
        class="chakra-text css-5km7a7"
      >
        TEST#1
      </p>
    </div>
    <button
      aria-label="Join platform button"
      class="css-y6mgpv"
    >
      <div
        class="css-tkevr6"
      >
        Join Platform!
      </div>
    </button>
  </div>
</div>
`;
